Python/Product/PythonTools/PythonTools/Repl/InteractiveWindowProvider.cs

```diff
@@ -91,13 +94,27 @@ class InteractiveWindowProvider {
             return false;
         }
 
-        public IVsInteractiveWindow Open(string replId) {
+        public void OnWindowUsed(IVsInteractiveWindow window) {
+            if (window == null) {
+                throw new ArgumentNullException(nameof(window));
+            }
+
+            lock (_windows) {
+                _lruWindows.Remove(window);
+                _lruWindows.Add(window);
+            }
+        }
+
+        public IVsInteractiveWindow Open(string replId) => Open(replId, null);
+
+        public IVsInteractiveWindow Open(string replId, Func<IInteractiveEvaluator, bool> predicate) {
```

Comment:
This code becomes very complicated. It has two callbacks, both happen under lock, and may return null even if window is already open. Method requires implicit information about the caller and caller has to know this method implementation, which pretty much devalues any decoupling between `InteractiveWindowProvider` and `ExecuteInReplCommand.EnsureReplWindow`. 

`IVsInteractiveWindow.Show` is publicly available, so `InteractiveWindowProvider` can simply return window instance and leave window opening to the `EnsureReplWindow` itself. Or the whole `EnsureReplWindow` can be moved into `InteractiveWindowProvider`, so that it knows what is actually going on.

Reply:
The lock is probably unnecessary, as you can't safely call this from off the UI thread anyway. I don't see the second callback (unless you mean `OnWindowUsed`? Which isn't actually a callback and could easily be inlined here if I wanted to duplicate the code twice).

All of the complexity already existed, BTW. By suggesting a refactoring for the sake of it you're increasing the complexity of this change.

Reply:
I'm totally fine with slightly bigger change, technical debts should be resolved as soon as they can be resolved. The small size of the change is actually not that important if we consider the amount of things one has to know and remember to understand it:

1. Input argument for predicate is always a `SelectableReplEvaluator`, otherwise it won't be called.
1. If `SelectableReplEvaluator.Evaluator` property is `null` or not a `PythonCommonInteractiveEvaluator` for project-matching `replId`, window it will be reused. Before the change, it was the opposite. The type doesn't really matter, because with current implementation it can be only `PythonInteractiveEvaluator` which is derived from `PythonCommonInteractiveEvaluator`, but property itself can be `null` if `SelectableReplEvaluator.SetEvaluator` has failed to create evaluator. So change is important, but not for the bug fix.
1. If `SelectableReplEvaluator._evaluator` is set, it won't change until `SelectableReplEvaluator` instance is disposed.
1. `AssociatedProjectHasChanged == true` doesn't mean that active project has changed, it means that interpreter for that project has changed or file structure has changed (project is assigned by `PythonReplEvaluatorProvider` and is never changed after that) and no action has been done . It also doesn't mean that REPL is stopped, hanged, closed, not yet initialized, etc. But it is set to `false` only by `PythonInteractiveEvaluator.EnsureConnectedAsync`, which is called when something is executed in REPL but not when it is actually reset.

So the bug was actually caused by the REPL window that had the lowest id in `InteractiveWindowProvider._windows` and was never used after interpreter has changed. Significant part of the fix is `_lruWindows` list that checks for newly created window before the old one, and `predicate` doesn't actually matter. Is that right?

Reply:
1. This is not guaranteed to always be the case. From outside the class, you can't tell this or rely on it.
2. I don't think there was a change here in the default case. If you don't provide a predicate, `null != false` is always true and so the condition is irrelevant.
3. Huh? It changes through `SetEvaluator`, not through `OpenOrCreate`. Python supports arbitrary numbers of interactive windows and each one has its own selector for the associated environment.
4. That is all it is meant to mean. You're right that resetting the window only resets that flag because of other hooked events - I'll add it to the `ResetAsync` calls.

You're right that in this particular case the logic works out such that the most recently used window is also the one we want to use (if any), so the predicate isn't necessary and we could simply check the result of `Open(replId)`. However, as soon as we make any other change that could reorder the windows, that breaks, so I'd rather have the more robust change in now so that we are protected.